File: InsurancePolicy/Architect.API.Insurance.DataAccess/Extender/Risk.vb
Imports Architect.Common.Proxy

Namespace Architect.API.Insurance.DataAccess.Policy

    Partial Public NotInheritable Class Risk


        Public Shared Function RetrieveLastIdByExecutiveUserCode(companyId As Integer, userId As Integer) As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT NVL(MAX(Id),0) " &
                                          "FROM Risk " &
                                         "WHERE CompanyId=@:CompanyId AND ExecutiveUserCode=@:UserId", "Risk", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                .AddParameter("UserId", DbType.Decimal, 5, False, userId)
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        Public Shared Function RetrieveLastPolicyId(companyId As Integer) As Integer
            Dim rows As Integer = 0

            With New DataManagerFactory("SELECT NVL(MAX(Policyid),0) " &
                                          "FROM Risk " &
                                         "WHERE CompanyId=@:CompanyId", "Risk", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .QueryExecuteScalarToInteger
            End With
            Return rows
        End Function

        Public Shared Function BuildFilter(filter As String, lineOfBusiness As Integer, product As Integer, status As String) As String
            Dim result As String = String.Empty

            'Permite filtrar por póliza, identificación, nombre, apellido
            'Email, telefono
            'Permite filtrar por certificado, identificación, número de cuenta
            If filter.IsNotEmpty Then
                If IsNumeric(filter) Then
                    result = " AND (P.POLICYID LIKE '%" & filter.ToUpper & "%' " &
                               "OR REPLACE(A.DOCUMENTNUMBER,'-','') LIKE '%" & filter.ToUpper & "%' " &
                               "OR A.PHONENUMBER LIKE '%" & filter.ToUpper & "%') "
                Else
                    result = " AND (UPPER(A.FIRSTNAME) LIKE '%" & filter.ToUpper & "%' " &
                               "OR UPPER(A.MIDDLENAME) LIKE '%" & filter.ToUpper & "%' " &
                               "OR UPPER(A.LASTNAME) LIKE '%" & filter.ToUpper & "%' " &
                               "OR UPPER(A.SECONDLASTNAME) LIKE '%" & filter.ToUpper & "%' " &
                               "OR A.DOCUMENTNUMBER LIKE '%" & filter.ToUpper & "%' " &
                               "OR UPPER(A.PRIMARYEMAILADDRESS) LIKE '%" & filter.ToUpper & "%') "
                End If

            End If

            If lineOfBusiness.IsNotEmpty Then
                result &= " AND P.LINEOFBUSINESSCODE=" & lineOfBusiness & " "
            End If
            If product.IsNotEmpty Then
                result &= " AND P.PRODUCTCODE=" & product & " "
            End If
            If status.IsNotEmpty Then
                result &= " AND P.STATUS IN (" & status & ") "
            End If

            Return result
        End Function

        ''' <summary>
        ''' Lista de pólizas existen según los criterio definidos en los filtros.
        ''' </summary>
        ''' <param name="companyId">Identificación de la compañia propietaria.</param>
        ''' <param name="filter">Permite filtrar por póliza o por la identificación, nombre, apellido, teléfono o correo electrónico del asegurado.</param>
        ''' <param name="lineOfBusiness">Opción para filtrar por ramo o linea de negocio.</param>
        ''' <param name="product">Opción para filtrar por producto vincaulado a un ramo.</param>
        ''' <param name="status">Opción para filtrar por multiples esta de las pólizas.</param>
        ''' <returns>Lista de pólizas según los criterio definidos en el filtro.</returns>
        Public Shared Function RetrieveView(companyId As Integer, filter As String, lineOfBusiness As Integer, product As Integer, status As String) As List(Of Contracts.Policy.PolicyView)
            Dim result As New List(Of Contracts.Policy.PolicyView)
            Dim selectDataTableItem As DataTable

            With New DataManagerFactory(My.Resources.Risk_RetrieveView.Replace("@FILTER@", BuildFilter(filter, lineOfBusiness, product, status)), "Risk", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                selectDataTableItem = .QueryExecuteToTable(True)
            End With
            If Not IsNothing(selectDataTableItem) AndAlso selectDataTableItem.Rows.Count > 0 Then
                Dim newItem As Contracts.Policy.PolicyView
                For Each itemDataRow As DataRow In selectDataTableItem.Rows
                    newItem = New Contracts.Policy.PolicyView With {
                        .Id = itemDataRow.IntegerValue("ID"),
                        .IssueDate = itemDataRow.DateTimeValue("ISSUEDATE"),
                        .LineOfBusiness = itemDataRow.StringValue("LINEOFBUSINESS"),
                        .Product = itemDataRow.StringValue("PRODUCT"),
                        .PolicyId = itemDataRow.StringValue("POLICYID"),
                        .InsuredName = itemDataRow.StringValue("INSURED"),
                        .InsuredNumber = itemDataRow.StringValue("DOCUMENTNUMBER"),
                        .Module = itemDataRow.StringValue("MODULE"),
                        .EffectiveDate = itemDataRow.DateTimeValue("EFFECTIVEDATE"),
                        .EndingDate = itemDataRow.DateTimeValue("ENDINGDATE"),
                        .Status = itemDataRow.StringValue("STATUS"),
                        .StatusCode = itemDataRow.NumericValue("STATUSCODE"),
                        .Currency = itemDataRow.StringValue("CURRENCY"),
                        .CancellationDate = itemDataRow.DateTimeValue("CANCELLATIONDATE"),
                        .ReasonForStatus = itemDataRow.StringValue("REASONFORSTATUS"),
                        .InsuredAmount = itemDataRow.NumericValue("INSUREDAMOUNT"),
                        .AnnualPremium = itemDataRow.NumericValue("ANNUALPREMIUM")
                    }
                    result.Add(newItem)
                Next
            End If

            Return result
        End Function

        Public Shared Function RetrieveByKey(id As Integer, companyid As Integer) As Architect.API.Insurance.Contracts.Policy.Risk
            Dim rows As Data.DataTable = Nothing
            Dim result As Architect.API.Insurance.Contracts.Policy.Risk = Nothing
            With New DataManagerFactory("SELECT Id, OriginType, LineOfBusinessCode, ProductCode, PolicyId, Currency, ModuleCode, PaymentFrequency, IssueDate, EffectiveDate, EndingDate, InsuredAmountBase, InsuredAmountComplement, InsuredAmount, AnnualPremium, MonthlyPremium, Surcharge, Status, CancellationDate, ReasonForStatus, Risk.CompanyId, Risk.CustomData, Risk.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Risk.UpdateDate, CertificateId, Comments, Annotation, Risk.Reference, Risk.BranchOffice, ExecutiveUserCode " &
                                          "FROM Risk LEFT JOIN UserMember um ON um.UserId = Risk.UpdateUserCode " &
                                         "WHERE Id=@:Id AND Risk.CompanyId=@:CompanyId", "Risk", "Research")
                .AddParameter("Id", DbType.Decimal, 9, False, id)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyid)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    result = DataRowMapper(rows.Rows(0))
                End If
            End With
            Return result
        End Function

        Public Shared Function RetrieveByDocumentNumber(companyid As Integer, lineOfBusinessCode As Integer, productCode As Integer, recordId As Integer, documentType As Integer, documentNumber As String) As Utilities.Contracts.LookUpValue
            Dim rows As Data.DataTable = Nothing
            Dim result As Utilities.Contracts.LookUpValue = Nothing

            With New DataManagerFactory("SELECT P.Status, LKS.Description StatusDesc " &
                                          "FROM RiskRoles JOIN Risk P ON P.Id=RiskRoles.PolicyId AND P.LineOfBusinessCode=@:LineOfBusinessCode AND P.ProductCode=@:ProductCode AND P.Status = 10 " &
                                          "LEFT JOIN LOOKUP LKS ON LKS.LOOKUPID = 2050 AND LKS.LANGUAGE=1 AND LKS.CODE = P.STATUS " &
                                         "WHERE P.CompanyId=@:CompanyId AND TYPE=2 AND DOCUMENTTYPE=@:DocumentType AND DOCUMENTNUMBER=@:DocumentNumber AND RiskRoles.PolicyId <> @:PolicyId ORDER BY P.ID DESC", "Risk", "Research")

                .AddParameter("LineOfBusinessCode", DbType.Decimal, 9, False, lineOfBusinessCode)
                .AddParameter("ProductCode", DbType.Decimal, 9, False, productCode)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyid)
                .AddParameter("DocumentType", DbType.Decimal, 8, False, documentType)
                .AddParameter("DocumentNumber", DbType.AnsiString, 20, False, documentNumber)
                .AddParameter("PolicyId", DbType.Decimal, 9, False, recordId)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count > 0 Then
                    result = New Utilities.Contracts.LookUpValue With {.Code = rows.FirstRow.NumericValue("Status"), .Description = rows.FirstRow.StringValue("StatusDesc")}
                End If
            End With
            Return result
        End Function

        ''' <summary>
        ''' Actualiza el campo de referencia para un registro en la tabla Risk por medio de su clave primaria.
        ''' </summary>
        ''' <returns>Cantidad de registros actualizados.</returns>
        Public Shared Function UpdateReference(companyId As Integer, Id As Integer, reference As String) As Integer
            Dim rows As Integer = 0
            With New DataManagerFactory("UPDATE Risk " &
                                           "SET Reference=@:Reference " &
                                         "WHERE Id=@:Id AND CompanyId=@:CompanyId", "Risk", "Research")
                .AddParameter("Reference", DbType.AnsiString, 36, False, reference)
                .AddParameter("Id", DbType.Decimal, 9, False, Id)
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyId)
                rows = .CommandExecute()
            End With
            Return rows
        End Function

        ''' <summary>
        ''' Recupera una lista de registros en la tabla Risk.
        ''' </summary>
        ''' <param name="companyid"></param>
        ''' <param name="status"></param>
        ''' <returns>Lista de id</returns>
        Public Shared Function RetrieveByStatus(companyid As Integer, status As Integer) As List(Of Utilities.Contracts.LookUpValue)
            Dim rows As Data.DataTable = Nothing
            Dim result As New List(Of Utilities.Contracts.LookUpValue)
            With New DataManagerFactory("SELECT Risk.Id, Risk.Reference " &
                                          "FROM Risk " &
                                          "JOIN RiskBayer ON RiskBayer.Id=Risk.Id AND NVL(RiskBayer.HasDigitalSignature,0)=0 " &
                                         "WHERE Risk.CompanyId=@:CompanyId AND Risk.Status=@:Status ", "Risk", "Research")
                .AddParameter("CompanyId", DbType.Decimal, 5, False, companyid)
                .AddParameter("Status", DbType.Decimal, 5, False, status)
                rows = .QueryExecuteToTable(True)
                If rows.IsNotEmpty AndAlso rows.Rows.Count <> 0 Then
                    For Each itemRow As DataRow In rows.Rows
                        result.Add(New Utilities.Contracts.LookUpValue() With {.Code = itemRow.NumericValue("Id").ToString(), .Description = itemRow.StringValue("Reference")})
                    Next
                End If
            End With
            Return result
        End Function

    End Class

End Namespace

File: InsurancePolicy/Architect.API.Insurance.DataAccess/Extender/Risk.cs
using Architect.DataFactory;
using Architect.Utilities.Extensions;
using System;
using System.Collections.Generic;
using System.Data;
using DbType = Architect.DataFactory.Enumerations.DbType;

namespace Architect.API.Insurance.DataAccess.Policy
{
    public partial class Risk
    {
        public static int RetrieveLastIdByExecutiveUserCode(int companyId, int userId, IDbConnection connection = null)
        {
            return (int)Database.Select(@"SELECT NVL(MAX(Id),0)
                                            FROM Risk
                                           WHERE CompanyId=:CompanyId AND ExecutiveUserCode=:UserId")
                             .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                             .AddParameter("UserId", DbType.Decimal, 5, userId)
                             .QueryScalar<Decimal>(connection, "Research");
        }

        public static int RetrieveLastPolicyId(int companyId, IDbConnection connection = null)
        {
            return (int)Database.Select(@"SELECT NVL(MAX(Policyid),0)
                                            FROM Risk
                                           WHERE CompanyId=:CompanyId")
                            .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                            .QueryScalar<Decimal>(connection, "Research");
        }

        public static string BuildFilter(string filter, int lineOfBusiness, int product, string status)
        {
            string result = string.Empty;

            // Permite filtrar por póliza, identificación, nombre, apellido
            // Email, telefono
            // Permite filtrar por certificado, identificación, número de cuenta
            if (filter.IsNotEmpty())
            {
                if (filter.IsNumeric())
                {
                    result = " AND (P.POLICYID LIKE '%" + filter.ToUpper() + "%' " + "OR REPLACE(A.DOCUMENTNUMBER,'-','') LIKE '%" + filter.ToUpper() + "%' " + "OR A.PHONENUMBER LIKE '%" + filter.ToUpper() + "%') ";
                }
                else
                {
                    result = " AND (UPPER(A.FIRSTNAME) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(A.MIDDLENAME) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(A.LASTNAME) LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(A.SECONDLASTNAME) LIKE '%" + filter.ToUpper() + "%' " + "OR A.DOCUMENTNUMBER LIKE '%" + filter.ToUpper() + "%' " + "OR UPPER(A.PRIMARYEMAILADDRESS) LIKE '%" + filter.ToUpper() + "%') ";
                }
            }

            if (lineOfBusiness.IsNotEmpty())
            {
                result += " AND P.LINEOFBUSINESSCODE=" + lineOfBusiness + " ";
            }

            if (product.IsNotEmpty())
            {
                result += " AND P.PRODUCTCODE=" + product + " ";
            }

            if (status.IsNotEmpty())
            {
                result += " AND P.STATUS IN (" + status + ") ";
            }

            return result;
        }

        /// <summary>
        /// Lista de pólizas existen según los criterio definidos en los filtros.
        /// </summary>
        /// <param name="companyId">Identificación de la compañía propietaria.</param>
        /// <param name="filter">Permite filtrar por póliza o por la identificación, nombre, apellido, teléfono o correo electrónico del asegurado.</param>
        /// <param name="lineOfBusiness">Opción para filtrar por ramo o linea de negocio.</param>
        /// <param name="product">Opción para filtrar por producto vincaulado a un ramo.</param>
        /// <param name="status">Opción para filtrar por multiples esta de las pólizas.</param>
        /// <returns>Lista de pólizas según los criterio definidos en el filtro.</returns>
        public static List<Contracts.Policy.PolicyView> RetrieveView(int companyId, string filter, int lineOfBusiness, int product, string status, IDbConnection connection = null)
        {
            var result = new List<Contracts.Policy.PolicyView>();

            Database.Select(Properties.Resources.Risk_RetrieveView.Replace("@FILTER@", BuildFilter(filter, lineOfBusiness, product, status)))
                       .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                       .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                       {
                           result.Add(DataReaderToPolicyView(reader));
                       }));
            return result;
        }

        public static Contracts.Policy.PolicyView DataReaderToPolicyView(System.Data.IDataReader reader, Contracts.Policy.PolicyView newItem = null)
        {
            newItem = new Contracts.Policy.PolicyView()
            {
                Id = reader.IntegerValue("ID"),
                IssueDate = reader.DateTimeValue("ISSUEDATE"),
                LineOfBusiness = reader.StringValue("LINEOFBUSINESS"),
                Product = reader.StringValue("PRODUCT"),
                PolicyId = reader.StringValue("POLICYID"),
                InsuredName = reader.StringValue("INSURED"),
                InsuredNumber = reader.StringValue("DOCUMENTNUMBER"),
                Module = reader.StringValue("MODULE"),
                EffectiveDate = reader.DateTimeValue("EFFECTIVEDATE"),
                EndingDate = reader.DateTimeValue("ENDINGDATE"),
                Status = reader.StringValue("STATUS"),
                StatusCode = (int)Math.Round(reader.NumericValue("STATUSCODE")),
                Currency = reader.StringValue("CURRENCY"),
                CancellationDate = reader.DateTimeValue("CANCELLATIONDATE"),
                ReasonForStatus = reader.StringValue("REASONFORSTATUS"),
                InsuredAmount = (double)reader.NumericValue("INSUREDAMOUNT"),
                AnnualPremium = (double)reader.NumericValue("ANNUALPREMIUM")
            };

            return newItem;
        }

        public static Contracts.Policy.Risk RetrieveByKey(int id, int companyId, IDbConnection connection = null)
        {
            Architect.API.Insurance.Contracts.Policy.Risk result = null;
            Database.Select("SELECT Id, Risk.CompanyId, OriginType, LineOfBusinessCode, ProductCode, PolicyId, Currency, ModuleCode, PaymentFrequency, IssueDate, EffectiveDate, EndingDate, InsuredAmountBase, InsuredAmountComplement, InsuredAmount, AnnualPremium, MonthlyPremium, Surcharge, Status, CancellationDate, ReasonForStatus, CertificateId, Comments, Annotation, Risk.Reference, Risk.BranchOffice, ExecutiveUserCode, Risk.CustomData, Risk.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Risk.UpdateDate " +
                              "FROM Risk LEFT JOIN UserMember um ON um.UserId = Risk.UpdateUserCode " +
                             "WHERE Risk.Id=:Id AND Risk.CompanyId=:CompanyId")
                        .AddParameter("Id", DbType.Decimal, 9, id)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result = DataReaderToRisk(reader);
                        }));
            return result;
        }

        public static Utilities.Contracts.LookUpValue RetrieveByDocumentNumber(int companyid, int lineOfBusinessCode, int productCode, int recordId, int documentType, string documentNumber, IDbConnection connection = null)
        {
            Utilities.Contracts.LookUpValue result = null;

            Database.Select(@"SELECT P.Status, LKS.Description StatusDesc
                               FROM RiskRoles JOIN Risk P ON P.Id=RiskRoles.PolicyId AND P.LineOfBusinessCode=:LineOfBusinessCode AND P.ProductCode=:ProductCode AND P.Status = 10
                               LEFT JOIN LOOKUP LKS ON LKS.LOOKUPID = 2050 AND LKS.LANGUAGE=1 AND LKS.CODE = P.STATUS
                             WHERE P.CompanyId=:CompanyId AND TYPE=2 AND DOCUMENTTYPE=:DocumentType AND DOCUMENTNUMBER=:DocumentNumber AND RiskRoles.PolicyId <> :PolicyId ORDER BY P.ID DESC")
                        .AddParameter("LineOfBusinessCode", DbType.Decimal, 9, lineOfBusinessCode)
                        .AddParameter("ProductCode", DbType.Decimal, 9, productCode)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyid)
                        .AddParameter("DocumentType", DbType.Decimal, 8, documentType)
                        .AddParameter("DocumentNumber", DbType.AnsiString, 20, documentNumber)
                        .AddParameter("PolicyId", DbType.Decimal, 9, recordId)
                       .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                       {
                           result = new Utilities.Contracts.LookUpValue() { Code = reader.NumericValue("Status").ToString(), Description = reader.StringValue("StatusDesc") };
                       }));

            return result;
        }

        /// <summary>
        /// Actualiza el campo de referencia para un registro en la tabla Risk por medio de su clave primaria.
        /// </summary>
        /// <returns>Cantidad de registros actualizados.</returns>
        public static int UpdateReference(int companyId, int Id, string reference, IDbConnection connection = null)
        {
            return (int)Database.Update(@"UPDATE Risk 
                                             SET Reference=:Reference 
                                           WHERE Id=:Id AND CompanyId=:CompanyId")
                            .AddParameter("Reference", DbType.AnsiString, 36, reference)
                            .AddParameter("Id", DbType.Decimal, 9, Id)
                            .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                            .Execute(connection, "Research");
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Risk.
        /// </summary>
        /// <param name="companyid"></param>
        /// <param name="status"></param>
        /// <returns>Lista de id</returns>
        public static List<Utilities.Contracts.LookUpValue> RetrieveByStatus(int companyid, int status, IDbConnection connection = null)
        {

            var result = new List<Utilities.Contracts.LookUpValue>();
            Database.Select(@"SELECT Risk.Id, Risk.Reference  
                                FROM Risk  
                                JOIN RiskBayer ON RiskBayer.Id=Risk.Id AND NVL(RiskBayer.HasDigitalSignature,0)=0  
                               WHERE Risk.CompanyId=:CompanyId AND Risk.Status=:Status ")
                    .AddParameter("CompanyId", DbType.Decimal, 5, companyid)
                    .AddParameter("Status", DbType.Decimal, 5, status)
                    .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                     {
                         result.Add(new Utilities.Contracts.LookUpValue() { Code = reader.NumericValue("Id").ToString(), Description = reader.StringValue("Reference") });
                     }));
            return result;
        }

        /// <summary>
        /// Recupera una lista de registros en la tabla Risk.
        /// </summary>
        /// <param name="id"></param>
        /// <param name="companyId"></param>
        /// <returns>Lista de instancias de Risk</returns>
        public static List<Contracts.Policy.Risk> RetrieveByIdCompanyId(int id, int companyId, IDbConnection connection = null)
        {
            var result = new List<Contracts.Policy.Risk>();
            Database.Select("SELECT Id, Risk.CompanyId, OriginType, LineOfBusinessCode, ProductCode, PolicyId, Currency, ModuleCode, PaymentFrequency, IssueDate, EffectiveDate, EndingDate, InsuredAmountBase, InsuredAmountComplement, InsuredAmount, AnnualPremium, MonthlyPremium, Surcharge, Status, CancellationDate, ReasonForStatus, CertificateId, Comments, Annotation, Risk.Reference, Risk.BranchOffice, ExecutiveUserCode, Risk.CustomData, Risk.UpdateUserCode, um.FirstName || ' ' || um.LastName AS UpdateUserName, Risk.UpdateDate " +
                              "FROM Risk LEFT JOIN UserMember um ON um.UserId = Risk.UpdateUserCode " +
                             "WHERE Risk.Id=:Id AND Risk.CompanyId=:CompanyId")
                        .AddParameter("Id", DbType.Decimal, 9, id)
                        .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                        .Query(connection, "Research", new Action<System.Data.IDataReader>((reader) =>
                        {
                            result.Add(DataReaderToRisk(reader));
                        }));
            return result;
        }

        /// <summary>
        /// Elimina un registro en la tabla Risk por medio de su clave primaria.
        /// </summary>
        /// <param name="id"></param>
        /// <param name="companyId"></param>
        /// <returns>Cantidad de registros eliminados.</returns>
        public static int DeleteByIdCompanyId(int id, int companyId, IDbConnection connection = null)
        {
            return Database.Delete(@"DELETE FROM Risk 
                                     WHERE Id=:Id AND CompanyId=:CompanyId")
                              .AddParameter("Id", DbType.Decimal, 9, id)
                              .AddParameter("CompanyId", DbType.Decimal, 5, companyId)
                              .Execute(connection, "Research");
        }
    }
}